### DISCUSSION TITLE
webpack-5 new Worker with typescript files behaviors unexpected

### DISCUSSION THREAD

**Discussion by fenyiwudian** [webpack/webpack#14093] [Q&A]
# Bug report

webpack-5's document says we can use web worker directly without using other loaders [here](https://webpack.js.org/guides/web-workers/) like:

```
new Worker(new URL('./worker.js', import.meta.url));
```
but for project wrote in typescript, this expression will be:
```
new Worker(new URL('./worker.ts', import.meta.url));
```

which does not function as expected.

**What is the current behavior?**

`dist/worker.ts` file is generated in dist instead of `dist/worker.js`, loading ts file as worker file will cause mine type error, although contents of `dist/worker.ts` are transpiled javascript codes.

**If the current behavior is a bug, please provide the steps to reproduce.**

as above metioned, one typescript file as worker, and another typescript file using it via new Worker. would reproduce.

**What is the expected behavior?**
generating `dist/worker.js` file.

**Other relevant information:**
webpack version:   5.49.0
Node.js version: 14.17.4
Operating System:  MacOS Big Sur 11.4
Additional tools: typescript@4.3.5 ts-loader@9.2.5

**Comment by alexander-akait** [comment]
Can you provide example of configuration?

**Reply by fenyiwudian** [reply]
Sorry, I made a mistake, It‘s not a bug。

actually
```
new Worker(new URL('./worker.ts', import.meta.url));
```
will work.

in my real case，I separated new Worker and new URL like：
```
const url = new URL('./worker.ts', import.meta.url);
new Worker(url);
```
because I have a worker pool function receive many worker urls and new many workers later, so I cannot new Worker and new URL at the same time。

later I realized in this case new URL was handle by normal `asset/resource` loader but not by worker specified `asset/resource` loader，
so I guess if I want to generate js from `new URL('./worker.ts', import.meta.url)`，I should provide `output.assetModuleFilename` to rename generated file, I tried and succeeded: 
```
 output: {
      filename: [name].js',
      assetModuleFilename: (pathData) => {
        if (pathData.filename.match(/worker\.ts$/)) {
          return '[name].js';
        } else {
          return 'assets/[name][ext]';
        }
      },
      path: path.resolve(__dirname, 'dist'),
    },
```
Thanks for your response and great work.
